--- /usr/share/initramfs-tools/scripts/functions.orig	2024-08-23 03:30:14.000000000 +0800
+++ /usr/share/initramfs-tools/scripts/functions	2024-11-24 09:53:50.988520800 +0800
@@ -433,6 +433,15 @@
 	LABEL=* | UUID=* | PARTLABEL=* | PARTUUID=*)
 		DEV="$(blkid -l -t "$DEV" -o device)" || return 1
 		;;
+###DO#NOT#DELETE#THIS#MARKER###
+	HIDDENPART=*)
+		local OFFSET=$(echo ${DEV#*=} | cut -d+ -f2)
+		local SIZELIMIT=$(echo ${DEV#*=} | cut -d+ -f3)
+		local DISK=$(lsblk -o NAME,PATH,PTUUID | grep $(echo ${DEV#*=} | cut -d+ -f1) | grep '^[a-z]' | awk '{print $2}')
+		losetup -D
+		DEV="$(losetup -f)"
+		losetup --offset $OFFSET --sizelimit $SIZELIMIT $DEV $DISK
+		;;
 	esac
 	[ -e "$DEV" ] && echo "$DEV"
 }
