--- /usr/share/initramfs-tools/init.orig	2021-01-15 02:04:11.000000000 +0800
+++ /usr/share/initramfs-tools/init	2021-12-25 17:35:52.002378204 +0800
@@ -68,7 +68,22 @@
 export fastboot=n
 export forcefsck=n
 export fsckfix=
-
+###DO#NOT#DELETE#THIS#MARKER###
+export KLOOP=
+export KROOT=
+export KLVM=
+export HOSTFSTYPE=
+export KLOOPFSTYPE=
+export SQUSHFS=
+export UPPERDIR=
+export LOWERDIR=
+export WORKDIR=
+export QEMUNBD=
+
+# mdadm needs hostname to be set. This has to be done before the udev rules are called!
+if [ -f "/etc/hostname" ]; then
+        /bin/hostname -F /etc/hostname >/dev/null 2>&1
+fi
 
 # Bring in the main config
 . /conf/initramfs.conf
@@ -104,6 +119,26 @@
 			;;
 		esac
 		;;
+	roottimeout=*)
+		ROOTDELAY="${x#roottimeout=}"
+		case ${ROOTDELAY} in
+		*[![:digit:].]*)
+			ROOTDELAY=
+			;;
+		esac
+		;;
+	loop=*)
+		# shellcheck disable=SC2034
+		LOOP="${x#loop=}"
+		;;
+	loopflags=*)
+		# shellcheck disable=SC2034
+		LOOPFLAGS="-o ${x#loopflags=}"
+		;;
+	loopfstype=*)
+		# shellcheck disable=SC2034
+		LOOPFSTYPE="${x#loopfstype=}"
+		;;
 	nfsroot=*)
 		# shellcheck disable=SC2034
 		NFSROOT="${x#nfsroot=}"
@@ -184,6 +219,36 @@
 	fsck.repair=no)
 		fsckfix=n
 		;;
+	kloop=*)
+		KLOOP=${x#kloop=}
+		;;
+	kroot=*)
+		KROOT=${x#kroot=}
+		;;
+	klvm=*)
+		KLVM=${x#klvm=}
+		;;
+	hostfstype=*)
+		HOSTFSTYPE=${x#hostfstype=}
+		;;
+	kloopfstype=*)
+		KLOOPFSTYPE=${x#kloopfstype=}
+		;;
+	squashfs=*)
+		SQUASHFS=${x#squashfs=}
+		;;
+	upperdir=*)
+		UPPERDIR=${x#upperdir=}
+		;;
+	lowerdir=*)
+		LOWERDIR=${x#lowerdir=}
+		;;
+	workdir=*)
+		WORKDIR=${x#workdir=}
+		;;
+	qemunbd=*)
+		QEMUNBD=${x#qemunbd=}
+		;;
 	esac
 done
 
@@ -198,7 +263,7 @@
 	resume=${RESUME:-}
 fi
 
-mount -t tmpfs -o "nodev,noexec,nosuid,size=${RUNSIZE:-10%},mode=0755" tmpfs /run
+mount -t tmpfs -o "nodev,noexec,size=50%,mode=0755" tmpfs /run
 mkdir -m 0700 /run/initramfs
 
 if [ -n "$log_output" ]; then
@@ -314,6 +379,16 @@
 unset forcefsck
 unset fsckfix
 unset starttime
+unset KLOOP=
+unset KROOT=
+unset KLVM=
+unset HOSTFSTYPE=
+unset KLOOPFSTYPE=
+unset SQUSHFS=
+unset UPPERDIR=
+unset LOWERDIR=
+unset WORKDIR=
+unset QEMUNBD=
 
 # Move virtual filesystems over to the real filesystem
 mount -n -o move /sys ${rootmnt}/sys
