--- /usr/share/initramfs-tools/init.orig	2024-11-14 21:32:26.000000000 +0800
+++ /usr/share/initramfs-tools/init	2024-11-24 20:37:40.538342415 +0800
@@ -82,7 +82,19 @@
 export fastboot=n
 export forcefsck=n
 export fsckfix=
-
+export recovery=
+###DO#NOT#DELETE#THIS#MARKER###
+export KLOOP=
+export KROOT=
+export KLVM=
+export HOSTFSTYPE=
+export KLOOPFSTYPE=
+export SQUSHFS=
+export UPPERDIR=
+export LOWERDIR=
+export WORKDIR=
+export QEMUNBD=
+export ROOTSUBVOL=
 
 # mdadm needs hostname to be set. This has to be done before the udev rules are called!
 if [ -f "/etc/hostname" ]; then
@@ -112,6 +124,9 @@
 	rootflags=*)
 		ROOTFLAGS="-o ${x#rootflags=}"
 		;;
+	rootsubvol=*)
+		ROOTSUBVOL="${x#rootsubvol=}"
+		;;
 	rootfstype=*)
 		ROOTFSTYPE="${x#rootfstype=}"
 		;;
@@ -232,6 +247,36 @@
 	fsck.repair=no)
 		fsckfix=n
 		;;
+	kloop=*)
+		KLOOP=${x#kloop=}
+		;;
+	kroot=*)
+		KROOT=${x#kroot=}
+		;;
+	klvm=*)
+		KLVM=${x#klvm=}
+		;;
+	hostfstype=*)
+		HOSTFSTYPE=${x#hostfstype=}
+		;;
+	kloopfstype=*)
+		KLOOPFSTYPE=${x#kloopfstype=}
+		;;
+	squashfs=*)
+		SQUASHFS=${x#squashfs=}
+		;;
+	upperdir=*)
+		UPPERDIR=${x#upperdir=}
+		;;
+	lowerdir=*)
+		LOWERDIR=${x#lowerdir=}
+		;;
+	workdir=*)
+		WORKDIR=${x#workdir=}
+		;;
+	qemunbd=*)
+		QEMUNBD=${x#qemunbd=}
+		;;
 	esac
 done
 
@@ -250,7 +295,7 @@
 	resume=${RESUME:-}
 fi
 
-mount -t tmpfs -o "nodev,noexec,nosuid,size=${RUNSIZE:-10%},mode=0755" tmpfs /run
+mount -t tmpfs -o "nodev,noexec,size=${RUNSIZE:-50%},mode=0755" tmpfs /run
 mkdir -m 0700 /run/initramfs
 
 if [ -n "$log_output" ]; then
@@ -376,6 +421,17 @@
 unset forcefsck
 unset fsckfix
 unset starttime
+unset KLOOP
+unset KROOT
+unset KLVM
+unset HOSTFSTYPE
+unset KLOOPFSTYPE
+unset SQUSHFS
+unset UPPERDIR
+unset LOWERDIR
+unset WORKDIR
+unset QEMUNBD
+unset ROOTSUBVOL
 
 # Move virtual filesystems over to the real filesystem
 mount -n -o move /sys ${rootmnt}/sys
