--- /usr/sbin/mkinitramfs.orig	2020-04-09 15:35:46.000000000 +0800
+++ /usr/sbin/mkinitramfs	2021-12-25 19:02:21.211611548 +0800
@@ -366,6 +366,20 @@
 # module-init-tools
 copy_exec /sbin/modprobe /sbin
 copy_exec /sbin/rmmod /sbin
+
+###DO#NOT#DELETE#THIS#MARKER###
+# util-linux
+copy_exec /sbin/blkid /sbin
+copy_exec /sbin/losetup /sbin
+copy_exec /sbin/kpartx /sbin
+copy_exec /bin/ntfs-3g /bin
+copy_exec /sbin/mount.exfat-fuse /bin
+copy_exec /sbin/vgscan /sbin
+copy_exec /sbin/shutdown /shutdown
+copy_exec /usr/bin/qemu-nbd /bin
+touch ${DESTDIR}/etc/initrd-release
+touch ${DESTDIR}/version
+
 mkdir -p "${DESTDIR}/etc/modprobe.d" "${DESTDIR}/lib/modprobe.d"
 for file in /etc/modprobe.d/*.conf /lib/modprobe.d/*.conf ; do
 	if test -e "$file" || test -L "$file" ; then
