--- /usr/sbin/mkinitramfs.orig	2023-01-17 10:22:57.000000000 +0800
+++ /usr/sbin/mkinitramfs	2024-08-05 14:36:10.102819564 +0800
@@ -244,6 +244,7 @@
 export MODULES
 export BUSYBOX
 export RESUME
+export FSTYPE
 
 # Private, used by 'catenate_cpiogz'.
 export __TMPCPIOGZ
@@ -300,6 +301,9 @@
 # Resolve hidden dependencies
 hidden_dep_add_modules
 
+# Add firmware for built-in code
+add_builtin_firmware
+
 # First file executed by linux
 cp -p /usr/share/initramfs-tools/init "${DESTDIR}/init"
 
@@ -345,6 +349,22 @@
 # module-init-tools
 copy_exec /sbin/modprobe /sbin
 copy_exec /sbin/rmmod /sbin
+
+###DO#NOT#DELETE#THIS#MARKER###
+# util-linux
+copy_exec /sbin/blkid /sbin
+copy_exec /sbin/losetup /sbin
+copy_exec /sbin/kpartx /sbin
+copy_exec /bin/ntfs-3g /bin
+copy_exec /sbin/mount.exfat-fuse /bin
+copy_exec /sbin/vgscan /sbin
+copy_exec /sbin/shutdown /shutdown
+copy_exec /usr/bin/qemu-nbd /bin
+copy_exec /usr/sbin/chkdsk /sbin
+copy_exec /usr/bin/lsblk /bin
+touch ${DESTDIR}/etc/initrd-release
+touch ${DESTDIR}/version
+
 mkdir -p "${DESTDIR}/etc/modprobe.d" "${DESTDIR}/lib/modprobe.d"
 for file in /etc/modprobe.d/*.conf /lib/modprobe.d/*.conf ; do
 	if test -e "$file" || test -L "$file" ; then
